google/cloud/pubsub/benchmarks/endurance.cc

```diff
@@ -0,0 +1,634 @@
+  void Shutdown();
+
+  std::vector<std::chrono::microseconds> ClearSamples() {
+    std::lock_guard<std::mutex> lk(mu_);
+    std::vector<std::chrono::microseconds> tmp;
+    tmp.swap(samples_);
+    return tmp;
+  }
+
+  void Debug(std::ostream& os) const {
+    std::lock_guard<std::mutex> lk(mu_);
+    os << "subscription_count=" << subscription_count_ << ", lwm=" << lwm_
+       << ", hwm=" << hwm_ << ", pending=" << pending_
+       << ", sent=" << sent_count_ << ", received=" << received_count_
+       << ", expected=" << expected_count_ << ", failures=" << failures_
+       << ", overflow=" << overflow_ << ", shutdown=" << shutdown_
+       << ", samples.size()=" << samples_.size();
+  }
+
+  std::int64_t SentCount() const {
+    std::lock_guard<std::mutex> lk(mu_);
+    return sent_count_;
+  }
+
+  std::int64_t ExpectedCount() const {
+    std::lock_guard<std::mutex> lk(mu_);
+    return expected_count_;
+  }
+
+  std::int64_t ReceivedCount() const {
+    std::lock_guard<std::mutex> lk(mu_);
+    return received_count_;
+  }
+
+  std::int64_t WaitReceivedCount(std::int64_t count) {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [&] { return received_count_ >= count; });
+    return received_count_;
+  }
+
+ private:
+  int const subscription_count_;
+  std::int64_t const lwm_;
+  std::int64_t const hwm_;
+
+  mutable std::mutex mu_;
+  std::condition_variable cv_;
+  std::int64_t pending_ = 0;
+  std::int64_t sent_count_ = 0;
+  std::int64_t expected_count_ = 0;
+  std::int64_t received_count_ = 0;
+  std::int64_t failures_ = 0;
+  bool overflow_ = false;
+  bool shutdown_ = false;
+  std::vector<std::chrono::microseconds> samples_;
+};
+
+bool ExperimentCompleted(Config const& config,
+                         ExperimentFlowControl const& flow_control,
+                         std::chrono::steady_clock::time_point start);
+
+void PublisherTask(Config const& config, ExperimentFlowControl& flow_control,
+                   int task);
+
+class Cleanup {
+ public:
+  Cleanup() = default;
+  ~Cleanup() {
+    for (auto i = actions_.rbegin(); i != actions_.rend(); ++i) (*i)();
+  }
+  void Defer(std::function<void()> f) { actions_.push_back(std::move(f)); }
+
+ private:
+  std::vector<std::function<void()>> actions_;
+};
+
+}  // namespace
+
+int main(int argc, char* argv[]) {
+  auto config = ParseArgs({argv, argv + argc});
+  if (!config) {
+    std::cerr << "Error parsing command-line arguments\n";
+    std::cerr << config.status() << "\n";
+    return 1;
+  }
+  if (config->show_help) return 0;
+
+  pubsub::TopicAdminClient topic_admin(pubsub::MakeTopicAdminConnection());
+  pubsub::SubscriptionAdminClient subscription_admin(
+      pubsub::MakeSubscriptionAdminConnection());
+
+  auto generator = google::cloud::internal::MakeDefaultPRNG();
+
+  auto const configured_topic = config->topic_id;
+
+  // If there is no pre-defined topic for this test, create one and
+  // automatically remove it at the end of the test.
```

Comment:
It looks like we delete the topic unconditionally.  Should be only do that if we created one?

Reply:
Good catch, fixed.